SEP 002 -- Add circular field to ComponentDefinition
====================================================

SEP                     | 
----------------------|--------------
**Title**                | Add circular field to ComponentDefinition
**Authors**           | Raik Gruenberg (raik.gruenberg at gmail com)
**Editor**            | Raik Gruenberg
**Type**               | Data Model
**SBOL Version** | 2.2
**Status**             | Draft
**Created**          | 03-Sep-2016
**Last modified**  | 

Abstract
-----------

SBOL currently is missing any means to distinguish between circular
(e.g. plasmid) and linear DNA constructs. A `circular` field at the level of
ComponentDefinition is proposed to remedy this problem.


1. Rationale <a name="rationale"></a>
----------------

Engineered DNA constructs are currently mostly deposited, used and exchanged as
circular plasmid DNA. However, there are also many applications for linear
constructs, which are easily and cheaply accessible through PCR or gene
synthesis. Moreover, even circular plasmid DNA may be cut and stored or
exchanged as linear DNA to aid the construction of new vectors. Likewise,
natural or engineered genomes may be circular (bacterial chromosomes) or linear
(eukaryotic chromosomes). 

Knowing whether a given construct is supposed to have circular or linear
topology is of crucial relevance both for experimental design and for basic
sequence analysis. Nevertheless, unlike the much older genbank format, SBOL so
far lacks the possibility to distinguish between linear and circular DNA.


2. Specification <a name="specification"></a>
----------------------------------------------

ComponentDefinition should be amended by a new field `circular` pointing to a
Boolean literal of either "True" or "False". This field will be *obligatory*, at
least, for ComponentDefinitions of type 'DNA'.

Circular `RNA` or `Protein` (in particular peptide) molecules are also
possible. Therefore, there is no need to restrict the use of the `circular`
field by component type. The distinction is crucial in case of
DNA. Given that the large majority of SBOL records currently describe DNA, the
field should be non-optional for ComponentDefinition in general.

In all cases, `circular=True` would instruct software to interpret the beginning
/ end position of a given sequence (be it DNA, amino acid or even smiles string)
as only arbitrary so that sequence features may be mapped or identified across
this junction.


3. Example or Use Case <a name='example'></a>
-------------------------------

This is such a basic feature of molecular biology that it is difficult to find
an example where it is, in fact, *not* needed. By way of example, consider
popular sequence design software such as Benchling, CLC, VectorNTI or others. In
these cases, declaring a sequence as circular changes the view and the searching
behaviour of the sequence editor.


4. Backwards Compatibility <a name='compatibility'></a>
-----------------

This is a new field and no backwards compatibility issues are foreseen. The field
will simply be missing from older records, in which case software should fall
back to making an educated guess or assuming linear topology.


5. Discussion <a name='discussion'></a>
-----------------

### 5.1 making `circular` an optional field

An alternative would be to declare `circular` as an *optional* field and to then
interpret the absence of the field as `circular = False`. However, absence of a
field should, generally speaking, not convey meaning. By contrast, absence of the `circular` field
would be against specification and signal true lack of information. Software or
human consumers may then be able to make an educated guess as to whether or not
a given construct is likely circular. For example, sequences mentioning
"plasmid" in their name should be assumed circular by default. For many other
parts, the default choice will likely be linear.

### 5.2 attaching `circular` to Component

An alternative would be to attach `circular` to `Component` rather than
`ComponentDefinition`. This would have the presumed advantage, that, for
example, a given plasmid record could be referred to both in linear and circular
state. This however seems not advisable. The linearized version of a plasmid is
fundamentally a different molecule, thus going beyond the "instantiation of
ComponentDefinition" use case of `Component`. Furthermore, any circular sequence
can be linearized in (length - 1) different ways and the actual position of
linearization (e.g. the actual beginning and end) should always be made explicit
through a dedicated `ComponentDefinition (circular=False)` record.

### 5.3 Formulation through sequence constraints

It would certainly be possible to formulate topology through sequence
constraints. In my opinion, this seems not advisable. Sequence constraints have
a different use case -- they describe incomplete designs that may, for example,
be communicated to services for automated design or library construction. This
is an advanced use case scenario that is not currently common. Standard sequence
editors will likely not support any sequence constraints framework in
the near future. By contrast, the current standard use case for SBOL is (or
should be) the communication of a complete sequence, e.g. for a plasmid
deposited at AddGene or a fragment described on parts.igem.org. Standard
sequence editors in general use do already distinguish between circular and
linear topology. This very basic feature should therefore be a dedicated field
at the level of ComponentDefinition for easy accessibility by such tools.


### 5.4 software implementation

In the case of a circular product, beginning and end position of the sequence
are entirely arbitrary. The same sequence can be wrapped (indexed) in many
different ways. This often leads to the apparent (but arbitrary) disruption of
sequence features. Software thus needs to be able to match and identify
sub-sequences or features that span the arbitrary beginning / end junction. By
contrast, even for a otherwise identical sequence, such a match should not be
reported if the construct is linear.


6. Competing SEPs <a name='competing_seps'></a>
-----------------

Currently none.


References <a name='references'></a>
----------------

[SBOL]: http://sbolstandard.org
[1]: https://www.python.org/dev/peps/pep-0001


Copyright <a name='copyright'></a>
-------------

<p xmlns:dct="http://purl.org/dc/terms/" xmlns:vcard="http://www.w3.org/2001/vcard-rdf/3.0#">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
  <br />
  To the extent possible under law,
  <a rel="dct:publisher"
     href="sbolstandard.org">
    <span property="dct:title">SBOL developers</span></a>
  have waived all copyright and related or neighboring rights to
  <span property="dct:title">SEP 008</span>.
This work is published from:
<span property="vcard:Country" datatype="dct:ISO3166"
      content="US" about="sbolstandard.org">
  United States</span>.
</p>
