SEP 004 -- Add contextual role properties to SequenceAnnotation and Component
===================================
SEP                   | 004
----------------------|--------------
**Title**             | Add contextual role properties to SequenceAnnotation and Component
**Authors**           | Mike Bissell (bissell at amyris.com), Raik Gruenberg (raik dot gruenberg at gmail), Chris Macklin (macklin at amyris.com)
**Editor**            | Bryan Bartley
**Type**              | Data Model
**SBOL Version**      | 2.1.0
**Replaces**          |
**Status**            | Draft
**Created**           | 10-Oct-2015
**Last modified**     | 16-Mar-2016

Abstract
--------

We propose adding optional `roles` fields to both `SequenceAnnotation` and
`Component`. A `roles` element comprises a set of zero or more `role` members,
which we call "contextual roles." Contextual roles will allow users to specify
the actual role(s) of a subcomponent or subsequence in the context of a design.

SBOL already permits roles on `ComponentDefinition` instances, however the
declared roles that were originally imagined and encoded by the author of a
`ComponentDefinition` may not accurately reflect the roles of that
part in every biological and operational context. In practice, context often
takes precedence in the determination of a part's actual roles. For example, a
DNA sequence encoding a protein purification tag might, in a particular design,
be used merely as a benign spacer between two DNA elements.


1. Rationale
------------

### BACKGROUND: ROLES PER SBOL 2.0

In [SBOL] 2.0, a non-contextual `role` field already belongs to
`ComponentDefinition`. This optional field classifies an entity's *declared*
function at the time of design. Roles *may* specify biological function, in
which case the roles *should* come from Table 3 (Sequence Ontology) if possible,
however [SBOL] does not strictly limit the application of roles to *biological*
ontologies. `role` simply links the given `ComponentDefinition` to a term from
an external ontology.


### ARGUMENT: CONTEXT MATTERS

Note, however, that in the context of a larger design, the *actual purpose* of
this `ComponentDefinition` may be rather different from that which was intended;
the original authors of a `ComponentDefinition` record cannot foresee all
possible applications of a biological part. In fact, outside of a host context,
a part in isolation most likely will fail to play its declared biological roles.
Moreover, genetic designers may want to annotate functions of a sequence that
are only locally relevant to their particular experiment or a specific
fabrication step. Consider, for example, sequences required in DNA assembly and
quality control, where the same DNA sequence could serve different purposes in
different locations of a design, or in different steps of an assembly workflow.

Importantly, when implementing a construction and testing service, or when
implementing genetic CAD software and the underlying genetic compilers, it
may become necessary to flag subcomponents in a design not only according to
their biological function, but also with non-biological classifications that
describe each subcomponent's logical role in each *specific assembly*.

During the automated planning of construction recipes, the intended biological
function of a physically isolated subcomponent is largely irrelevant, but it
may become necessary to classify logistical and compositional relationships in
order to manage the flow of templates, host strains, construction reagents,
and output products through a construction workflow. For example, a planning
algorithm might decide that one `Component` incorporating a PCR primer's
`ComponentDefinition` is to be implemented by ordering an oligo from a
particular vendor, while a second such `Component` is to be implemented by
retrieving an existing sample from inventory.

During computer aided design and compilation, certain algorithms may care about
the local, logical or compositional role of an entity while deliberately
ignoring its planned biological function in finished assemblies. Since modular
subcomponents are typically intended for reuse in multiple finished assemblies,
these subcomponents' eventual logical or biological roles may not even be
determined at the time of design or compilation, however the genetic compiler
will still need to classify subcomponents' relationships to various intermediate
build products containing them.

In both scenarios, non-biological roles are important because a subcomponent has
no relevant biological activity in isolation, during design and construction.
Therefore, contextual roles are needed.


### ARGUMENT FOR BOTH COMPONENT ROLES AND SEQUENCEANNOTATION ROLES

We propose that contextual roles are required in two places:
`Component` and `SequenceAnnotation`.

Adding an optional `roles` field to `Component` will allow users to describe the
subcomponent's actual design purpose with respect to the current assembly,
without reference to the specific details of its parent's sequence (if that
sequence is even known yet).

Adding an optional `roles` field to `SequenceAnnotation` will allow users to
describe the design purpose of a subsequence and an optional subcomponent in the
context of a parent sequence, with respect to a specific location in that
sequence.

It is insufficient to put roles only on `SequenceAnnotation`; roles are
additionally needed on `Component`. There are two reasons. First, not all roles
pertain to a specific parent sequence, or to a particular location on that
sequence. Second, SequenceAnnotations *must* refer to a sequence, but a parent
sequence is not always known at every moment in the lifetime of a design. For
instance, genetic compilers often compute the sequences of a parent assembly
from the sequences of its children in a bottom-up fashion. In this case, the
parent sequence would not be determined until the final compilation step, even
though the contextual roles of its subcomponents were determined earlier in the
build. Compilation of the parent ComponentDefinition's sequence might even be
deferred until later, when the host context is finally determined. Between the
time that the subcomponents were compiled and the moment of the parent's
compilation, subcomponents' contextual roles must be conveyed through the
compiler toolchain, and we propose that `Component.roles` is the appropriate
location.

It is insufficient to put roles only on `Component`; roles are additionally
needed on `SequenceAnnotation`. There are, again, two reasons. First, when it is
possible to map a subcomponent with some contextual `role` (say, the promoter
http://identifiers.org/so/SO:0000167) onto a *specific location* in a parent
sequence, then users will want to do that. Only `SequenceAnnotation` gives us
that ability; `Component` does not refer to parent sequence details. Secondly,
users may choose not to model every aspect of a `ComponentDefinition`
hierarchically (using Components to link in sub-ComponentDefinitions), but our
users may still wish to annotate the genetic details contained in a
ComponentDefinition's sequence. For example, the user might tag one subsequence
as an RBS, and another sequence as a start codon, without making separate,
heavyweight `Component` and `ComponentDefinition` elements for the RBS and the
start codon.


2. Specification
----------------

We specify the contents of `SequenceAnnotation.roles` and `Component.roles` in
analogy to `Participation.role` ([SBOL] 7.9.4) and `ComponentDefinition.role`
([SBOL] 7.1):

### 2.1 Component.roles

The `Component.roles` property is an OPTIONAL set of zero or more `role` URIs
describing the purpose or potential function of a given child
`ComponentDefinition` in the context of its parent `ComponentDefinition`.

The data model distinguishes between an *empty* set of `Component.roles` and an
*absent* one.

If provided, the `roles` property MUST contain zero or more `role` URIs that
MUST identify terms from appropriate ontologies. Roles are not restricted to
describing biological function; they may annotate Components' function in any
domain for which an ontology exists.

It is RECOMMENDED that these `role` URIs identify terms that are compatible with
the `type` properties of both the parent and child ComponentDefinitions. For
example, a `role` of a `Component` which belongs to a `ComponentDefinition` of
type DNA and points to a child `ComponentDefinition` of type DNA might refer to
terms from the Sequence Ontology. A table of recommended ontology branches is
given in the [SBOL] specification.

Higher-level roles take precedence. If a `Component` has a set of zero or more
`roles` and its child `ComponentDefinition` also has a set of one or more `role`
elements, then the `Component.roles` set completely overrides all roles on its
child `ComponentDefinition` with respect to the current parent
`ComponentDefinition` context. Tools must not merge the potentially conflicting
contents of the two sets.

It is RECOMMENDED to specify a set of `Component.roles` only if that set of
roles differs from the set of `role` elements already belonging to its child
`ComponentDefinition`.

If a Component's set of `roles` contain multiple `role` URIs, then they MUST
identify non-conflicting terms (refer to examples in the [SBOL] spec).

### 2.2 SequenceAnnotation.roles

The `SequenceAnnotation.roles` property is an OPTIONAL set of zero or more
`role` URIs describing the purpose or potential function of a given subsequence
(and, if given, a child `ComponentDefinition`) in the context of its parent
`ComponentDefinition`.

The data model distinguishes between an *empty* set of
`SequenceAnnotation.roles` and an *absent* one.

If provided, the `roles` property MUST encode zero or more `role` URIs that MUST
identify terms from appropriate ontologies. Roles are not restricted to
describing biological function; they may annotate Sequences' function in any
domain for which an ontology exists.

It is RECOMMENDED that these `role` URIs identify terms that are compatible with
the `type` properties of both the parent and child ComponentDefinitions (if
given). For example, a `role` of a `SequenceAnnotation` which belongs to a
`ComponentDefinition` of type DNA and incorporates a child
`ComponentDefinition` of type DNA might refer to terms from the Sequence
Ontology. A table of recommended ontology branches is given in the [SBOL]
specification.

Higher-level roles take precedence. If a `SequenceAnnotation` has a set of zero
or more roles and either its optional sibling `Component` and/or the child
`ComponentDefinition` also has its own set of one or more roles, then the
`SequenceAnnotation.roles` set completely overrides all roles on its sibling
`Component` and the child `ComponentDefinition` with respect to the region
of the parent `Sequence` annotated by that `SequenceAnnotation`. Tools must not
merge the potentially conflicting contents of the three sets.

It is RECOMMENDED to specify a set of `SequenceAnnotation.roles` only if that
set of roles differs from the set of `roles` already belonging to its optional
sibling `Component` (if any). If a sibling `Component` exists but it has no
declared set of `Component.roles`, then it is RECOMMENDED to specify a set of
`SequenceAnnotation.roles` only if that set differs from the set of `role`
elements already belonging to the child `ComponentDefinition` (if any).

If a set of `SequenceAnnotation.roles` contain multiple `role` URIs, then they
MUST identify non-conflicting terms (refer to examples in the [SBOL] spec).

### 2.3 XML Serialization Details

For both proposed contextual `roles` properties, a set of zero or more `<role>`
child elements, if declared, MUST be contained within a single `<roles>` element
in the XML serialized form.

To declare that a `Component` or `SequenceAnnotation` has zero roles in the
current context, encode that empty set with a childless `<roles`> element. To
refrain from specifying a set of roles, include no `<roles>` element at all.

N.B. An empty `<roles>` element is not equivalent to an absent one. An empty
`<roles>` element positively implies an empty set of contextual roles for the
purposes of applying the precedence rules in subsections 2.1 and 2.2.

By providing a way to override a *nonempty* set of child roles with an *empty*
set of parent roles, this data structure allows us to specify that a child
`ComponentDefinition` whose designer declared some role(s) actually lacks those
role(s) in the current context.

### 2.4 Editorial remark

The language above about "non-conflicting terms" is boilerplate extracted from
the current [SBOL] 2.0 spec. It is evidently vague and open-ended. Note that the
[SBOL] spec remains silent on the topic of how to interpret "non-conflicting,"
because this will depend on the interpretation of these roles, a matter which
lies well outside the scope of [SBOL], as noted under `ModuleDefinition`:
"Interpretation of the meaning of such roles currently depends on the software
tools that read and write them." ([SBOL]) Given this vagueness, someone might
wish to propose in a separate SEP to eliminate this boilerplate, as it provides
no specific guidance to SBOL software developers.


3. Examples
-----------

### 3.1 DNA assembly and cloning

A collection of DNA constructs for which DNA assembly is ordered from a
commercial provider may contain a short DNA sequence in different locations of
several constructs. In some constructs, this sequence may be included as a
forward sequencing primer binding site, in others as a reverse primer for the
amplification of a DNA segment during the DNA assembly process.

### 3.2 Protein peptide sequences

A given fusion protein design may contain a Hexahistidine tag which is commonly
used for protein purification. However, in a cell biology experiment, the His
tag may instead have been included as an antigen in order to localize the
protein with immunostaining. In both cases, the `SequenceAnnotation` should
(via `Component`) point to the same `ComponentDefinition` (His tag). As long as
the actual purpose is annotated in `SequenceAnnotation`, a third party could
automatically verify whether this immunotag is compatible with the antibodies
available in-house. Alternatively, a cell biologist may automatically remove all
protein purification tags from a given design but would want to keep any
sequence used for immunostaining.


4. Backwards Compatibility
--------------------------

There are no obvious backwards compatibility issues. Contextual roles require
additional fields on both `Component` and `SequenceAnnotation`. These fields
can be safely ignored by software tools written for [SBOL] 2.0.


5. Discussion
-------------

This proposal was advanced by Mike Bissell from Amyris in order to accommodate
two types of annotations that proved necessary to implement the Amyris genetic
compilers and DNA assembly processes.


6. Competing SEPs
-----------------

There are currently no competing or conflicting SEPs.


References
----------

SBOL - http://sbolstandard.org/downloads/specification-data-model-2-0/
[SBOL]: http://sbolstandard.org/downloads/specification-data-model-2-0/


Copyright
---------

This document has been placed in the public domain.
