SEP 010 -- simplify description of sequence features and sub-parts
===================================

SEP                   | 10
----------------------|--------------
**Title**             | simplify description of sequence features and sub-parts
**Authors**           | Raik Gruenberg <raik.gruenberg at gmail com>
**Editor**            | James McLaughlin
**Type**              | Data Model
**SBOL Version**      | 2.2, 3.0
**Replaces**          | 
**Status**            | Draft
**Created**           | 20-Sep-2016
**Last modified**     | 23-Sep-2016

Abstract
-----------

There are two very different types of 'part annotation'. (1) part composition relationships -- These always point to an existing (and presumably re-usable) sub-component. Sequence location or indeed sequence information may or may not be available. (2) Classic sequence feature annotations -- As known from the genbank format, these only apply to clearly specified sequence regions but often do not point to meaningful sub-parts.

Currently, `Component` alone is sufficient to describe sub-part relations without any sequence information. This however is the exception in synthetic biology practice. Both `SequenceAnnotation` and `Component` are needed for SBOL encoding of actual genetic designs with parts and sub-parts (because `Component` lacks a `location` field). Conversely, simple sequence features can be described using `SequenceAnnotation` alone (as of SBOL 2.0) but this possibility is not widely known and additional `Component` and `ComponentDefinition` are often created instead. 

We propose to modify `Component` and `SequenceAnnotation` such that `Component` is solely responsible for the description of part - subpart relationships (with or without sequence) and `SequenceAnnotation` is solely responsible for the description of genbank-style sequence features. `SequenceAnnotation` should subsequently (as of SBOL 3.0) be renamed to `SequenceFeature`.

Table of Contents 
---------------------

* [1. Rationale](#rationale)
  * 1.1 current situation
  * 1.2 Goals of the proposal
* [2. Specification](#specification)
  * 2.1 Add `location` field to `Component`
  * 2.2 Restrict `SequenceAnnotation` to sequence feature annotation
  * 2.3 Let `SequenceConstraint.object` point to `SequenceFeature`
  * 2.4 Rename `SequenceAnnotation` to `SequenceFeature`
* [3. Example or Use Case](#example)
* [4. Backwards Compatibility](#compatibility)
  * 4.1 suggested transition path
  * 4.2 Conversion of SBOL 2.1 records to 3.0
  * 4.3 Backwards conversion of SBOL 3.x to 2.1 records
* [5. Discussion](#discussion)
* [6. Competing SEPs](#competing_seps)
* [References](#references)
* [Copyright](#copyright)


1. Rationale <a name="rationale"></a>
----------------

### 1.1 current situation

The current `SequenceAnnotation` class has a dual purpose:

(1) Its primary role is to specify the location of "sub-parts" within the sequence of a parent ComponentDefinition. To this end, `SequenceAnnotation` links one or more `Location` records with a `Component`. The `Component`, in turn, refers to a `ComponentDefinition` (via its `definition` field). This `ComponentDefinition` is the description of the actual sub-part. Actual physical composition is therefore defined like this:

`ComponentDefinition` -[`sequenceAnnotation`]-> `SequenceAnnotation` -[`component`]-> `ComponentDefinition` -[`definition`]-> `ComponentDefinition`

The directionality (which one is parent and which one is a sub-part) is frequently confused. Moreover, the parent `ComponentDefinition` also directly links to the sub-part `Component` via a `component` field. This is necessary so that composition can be described before any sequences (and thus sequence locations) are known. An additional chain of references is therefore needed, in parallel to the one shown above:

`ComponentDefintion` -[`component`]-> `Component` -[`definition`]-> `ComponentDefinition`

Adding to this redundancy, both `Component` and `SequenceAnnotation` may have `role` properties that diverge from the `role` (functional classification) of the target `ComponentDefinition`. Whether a diverging `role` is attached to `Component` or `SequenceAnnotation` is an arbitrary choice. This invites conflicting implementations and interpretations of this field. 

Evidently, SBOL makes the description of "undefined", "loose bag", part composition without any sequence information relatively easy. By contrast, the description of actual genetic designs with actual sequences is surprisingly complex and redundant. This is unfortunate because the latter is, by and far, the overwhelming use case of SBOL. It also hinders adoption by sequence-level designers and tool developers. 

(2) The secondary role of `SequenceAnnotation` is to simply annotate regions of interest within a given sequence. Arguably, this should be its primary role (hence the name) as it is a very common use case in practice. A `SequenceAnnotation` without `component` **can** be created and linked to a region of, e.g., DNA. `SequenceAnnotation` inherits `name` and `description` fields from `Identified` and is therefore sufficient for the description of "flat" sequence features. 


### 1.2 Goals of the proposal

![proposed simplified data model][data model]

(1) Restrict the use of `SequenceAnnotation` to annotations of features which do not fall into the part - subpart category. As a welcome side effect, this should make it much easier to move back and forth between SBOL and large bodies of existing genbank-formatted information and related software.

(2) Simplify the part-subpart relationship via `Component` so that it does not any longer require `SequenceAnnotation`.

(3) Create a syntactic parallel between sequence/physical and functional part-subpart relations in SBOL -- The `Component` class will be equivalent in syntax and meaning to the existing `Participation` class. For programmers, the pattern `ComponentDefinition -> Component(role) -> ComponentDefinition` will look and feel like the already established pattern `Interaction -> Participation(role) -> ComponentDefinition`. 

(4) Remove ambiguity as to how things can / should be expressed at the sequence layer to aid meaningful data exchange. 



2. Specification <a name="specification"></a>
----------------------------------------------

### 2.1 Add `location` field to `Component`

Add the following *optional* field to `Component`:

  * [0..n] `location` pointing to a `Location` on the *parent* `ComponentDefinition` sequence; if `location` is missing, this indicates a part / sub-part relationship for which sequence details have not (yet) been determined. [SBOL 2.2]


### 2.2 Restrict `SequenceAnnotation` to sequence feature annotation

Remove the following fields from `SequenceAnnotation`:

  * `component` -- SequenceAnnotation is not any longer used for part - subpart relations
  * `roleIntegration` -- there is no sub-part/definition that `role` fields may be in conflict with

Fields will be deprecated as of SBOL 2.2 and removed in SBOL 3.0.

Update the specification to clarify usage of existing fields [SBOL 2.2]:

  * [0..n] `role` pointing to a SequenceOntology term (optional), corresponds to genbank type field
  * [0..1] `name` corresponding to genbank name field (optional but now RECOMMENDED)
  * [0..1] `description` corresponds to genbank description field (optional)

Moreover, a *validation rule* is needed: `SequenceAnnotation` can only be used
if an actual sequence record is specified for the parent `ComponentDefinition` [SBOL 3.0].

### 2.3 Let `SequenceConstraint.object` point to `SequenceAnnotation`

`SequenceConstraint.object` can point to either `ComponentInstance` derivatives
(as before) **or** to `SequenceAnnotation` (`SequenceFeature`). [SBOL 2.2]

This change allows to anchor constraints on sequence regions that are not
actually sub-parts. Examples may be start / stop codons, transcription start
sites or specific mutations.  `SequenceAnnotation` MUST NOT be `subject` of a
constraint.

### 2.4 Rename `SequenceAnnotation` to `SequenceFeature`

  * rename class `SequenceAnnotation` to `SequenceFeature`
  * rename `sequenceAnnotation` field of `ComponentDefinition` to `sequenceFeature`
  
This backwards-incompatible change is to be introduced with SBOL 3.0.


3. Example or Use Case <a name='example'></a>
-------------------------------

Example use cases for the modified SequenceAnnotation are feature annotations such as START or STOP codons, mutations, highlighting regions referred to in a paper, sequence conflicts, etc, all mainly intended for human consumption. Over the evolution of a design, sequence features may later be formalized into re-usable subparts (i.e. 'Component's) It is therefore conceivable that a sequence editor reads in a genbank file with many sequence features and offers the user the easy conversion of some of those features into sub-parts. This, in fact, is a workflow already used and supported by the Benchling Sequence editor (http://benchling.com).


4. Backwards Compatibility <a name='compatibility'></a>
-----------------

### 4.1 suggested transition path

Phase 1: SBOL 2.2
  * add `location` field to `Component`
  * mark `component` and `roleIntegration` as "deprecated" in `SequenceAnnotation`
  * advise software to stop using chains of SequenceAnnotation and Component in their SBOL output
  * allow `SequenceConstraint.object` to point to `SequenceAnnotation`

Phase 2: SBOL 3.0
  * remove SequenceAnnotation.component, SequenceAnnotation.roleIntegration
  * rename `SequenceAnnotation` to `SequenceFeature`
  * validation rule: `SequenceFeature` is only used with actual sequence attached to `ComponentDefinition`


### 4.2 Conversion of SBOL 2.1 records to 3.0:

  1. remove intermediate `SequenceAnnotation` and move `SequenceAnnotation.location` to `Component`
  2. optionally, try to flatten `SequenceAnnotation` - `Component` - `ComponentDefinition` chains of trivial annotations into `SequenceAnnotation` records

### 4.3 Backwards conversion of SBOL 3.x to 2.1:

  * conversion of localized `Component`:
  
     (1) create `SequenceAnnotation` record pointing to subpart `Component`

     (2) move `location` from `Component` to `SequenceAnnotation`

  * conversion of non-localized `Component`:
  
     no change required

  * conversion of `SequenceFeature`
  
     (1) rename `SequenceFeature` to `SequenceAnnotation`

     (2) rename `ComponentDefinition.sequenceFeature` field to `sequenceAnnotation`


5. Discussion <a name='discussion'></a>
-----------------

### 5.1 author comments

As an added benefit, the proposed change creates a symmetry between the sequence
and the functional layer of SBOL: `Component` is now the equivalent of
`Participation`. The former describes a physical part- subpart relation whereas
the latter describes a functional part - subpart relation. Both specify one or
more `role` properties, both point to a (sub)`ComponentDefinition`. Currently,
this parallel is obfuscated by the multiple direct and indirect references
between parent and sub-part `ComponentDefinition`.

### 5.2 discussion at COMBINE

  * It was pointed out that `SequenceAnnotation` already can have its own `name` and `description` fields as it is derrived from `Identified`. The SEP was changed accordingly.
  * Renaming `SequenceAnnotation` to `SequenceFeature` was universally considered a good idea (for symmetry with genbank, bioinformatics practics and in order to avoid confusion with "Annotation" in SBOL and SBML).
  * It was suggested to rename `SequenceConstraint` into `ComponentConstraint` -- should this be added to the SEP?
  * It was suggested that `SequenceConstraint` should also be allowed to point to `SequenceAnnotation`. This would avoid construction of `Component` - `ComponentDefinition` chains for, e.g. mutations or other simple features that are not sub-parts but nevertheless restrict/orient the positioning of other Components. This change has been incorporated into the SEP.


6. Competing SEPs <a name='competing_seps'></a>
-----------------

None.

References <a name='references'></a>
----------------

[SBOL]: http://sbolstandard.org
[1]: https://www.python.org/dev/peps/pep-0001
[data model]: https://github.com/SynBioDex/SEPs/blob/master/images/sep_010_model.png "Proposed simplified data model"

< refer to these references in the text as, e.g. [SBOL] or [1] >

Copyright <a name='copyright'></a>
-------------

<p xmlns:dct="http://purl.org/dc/terms/" xmlns:vcard="http://www.w3.org/2001/vcard-rdf/3.0#">
  <a rel="license"
     href="http://creativecommons.org/publicdomain/zero/1.0/">
    <img src="http://i.creativecommons.org/p/zero/1.0/88x31.png" style="border-style: none;" alt="CC0" />
  </a>
  <br />
  To the extent possible under law,
  <a rel="dct:publisher"
     href="sbolstandard.org">
    <span property="dct:title">SBOL developers</span></a>
  has waived all copyright and related or neighboring rights to
  <span property="dct:title">SEP 010</span>.
This work is published from:
<span property="vcard:Country" datatype="dct:ISO3166"
      content="US" about="sbolstandard.org">
  United States</span>.
</p>
 